core/src/main/java/org/eclipse/hono/util/HonoEventMultiplexer.java

```diff
@@ -0,0 +1,143 @@
+     * @return The event consumer that invokes the specific handlers depending on the event type.
+     * @throws NullPointerException If the notificationConsumer is {@code null}.
+     */
+    public static final Consumer<Message> createEventHandler(final Consumer<Message> eventConsumer,
+                                                             final Consumer<Message> notificationConsumer
+    ) {
+        Objects.requireNonNull(notificationConsumer);
+
+        return createEventHandler(eventConsumer, notificationConsumer, null);
+    }
+
+    /**
+     * Construct a event consumer that delegates different kinds of events to specific handlers.
+     *
+     * @param eventConsumer Consumer that handles generic events.
+     * @param notificationConsumer Consumer that handles events that were detected to be a notification. If {@code null}
+     *                             the consumer will be ignored.
+     * @param deviceReadyForCommandConsumer Consumer that is invoked if a notification was received that signals the command
+     *                                      readiness of a device. Note that when this consumer is invoked the notification
+     *                                      was already validated. If {@code null} the consumer will be ignored.
+     * @return The event consumer that invokes the specific handlers depending on the event type.
+     */
+    public static final Consumer<Message> createEventHandler(final Consumer<Message> eventConsumer,
+                                                             final Consumer<Message> notificationConsumer,
+                                                             final Consumer<NotificationDeviceCommandReadyObject> deviceReadyForCommandConsumer
+                                                             ) {
+        return (msg -> {
+            final Section body = msg.getBody();
+            if (!(body instanceof Data)) {
+                return;
+            }
+
+            if (HonoDeviceNotificationConstants.isNotification(msg)) {
+                // always call the general notification consumer, if provided
+                Optional.ofNullable(notificationConsumer).map(consumer -> {
+                    notificationConsumer.accept(msg);
+                    return consumer;
+                });
+
+                Optional.ofNullable(deviceReadyForCommandConsumer).map(consumer -> {
+                    // investigate if we need to call device ready for command consumer
+
+                    if (HonoDeviceNotificationConstants.CONTENT_TYPE_DEVICE_NOTIFICATION.equals(msg.getContentType())) {
+                        Optional.ofNullable(((Data)msg.getBody()).getValue()).
+                                map(data -> {
+                                    try {
+                                        return OBJECT_MAPPER.readValue(data.toString(), NotificationDeviceCommandReadyObject.class);
```

Comment:
in order to find out if this is an event indicating that a device is ready to receive a command, we need to try to deserialize it?

Reply:
Yes and No:
- if the event is matching at all is done by checking the content-type.
- if a matching event is valid regarding the contained data is done by looking into the JSON payload, which is done here by deserializing it. Alternative: parse it manually.

To make discussion easier, I will split this PR into
- one that contains only the definitions necessary for these notifications
- one that offers the object based functionality, so not every application needs to implement the parsing of a notification event itself.

I guess this will make the functionality contained here clearer, and we can have a quicker agreement regarding the definitions of the notification definitions.